shared/src/types/vote_extensions/validator_set_update.rs

```diff
@@ -0,0 +1,297 @@
+            Token::String("updateValidatorsSet".into()),
+            Token::FixedBytes(
+                ext.voting_powers.get_bridge_hash(ext.epoch).to_vec(),
+            ),
+            Token::FixedBytes(
+                ext.voting_powers.get_governance_hash(ext.epoch).to_vec(),
+            ),
+            epoch_to_token(ext.epoch),
+        ])
+    }
+
+    /// Sign this [`Vext`] with an Ethereum key.
+    ///
+    /// For more information, check the Ethereum bridge smart contract code:
+    ///   - <https://github.com/anoma/ethereum-bridge/blob/main/contracts/contract/Bridge.sol#L186>
+    pub fn new_abi_encoded(keypair: &common::SecretKey, ext: Vext) -> Self {
+        let to_sign = Self::serialize_vext(&ext);
+        let sig = common::SigScheme::sign(keypair, &to_sign);
+        Self::new_from(ext, sig)
+    }
+
+    /// Verify the signature of a [`Vext`], signed by some
+    /// Ethereum key.
+    pub fn verify_abi_encoded(
+        &self,
+        pk: &common::PublicKey,
+    ) -> Result<(), VerifySigError> {
+        let bytes = Self::serialize_vext(&self.data);
+        common::SigScheme::verify_signature_raw(pk, &bytes, &self.sig)
+    }
+}
+
+/// Represents a validator set update, for some new [`Epoch`].
+#[derive(
+    Eq, PartialEq, Clone, Debug, BorshSerialize, BorshDeserialize, BorshSchema,
+)]
+pub struct Vext {
+    /// The addresses of the validators in the new [`Epoch`],
+    /// and their respective voting power.
+    ///
+    /// When signing a [`Vext`], this [`VotingPowersMap`] is converted
+    /// into two arrays: one for its keys, and another for its
+    /// values. The arrays are sorted in descending order based
+    /// on the voting power of each validator.
+    pub voting_powers: VotingPowersMap,
+    /// TODO: the validator's address is temporarily being included
+    /// until we're able to map a Tendermint address to a validator
+    /// address (see https://github.com/anoma/namada/issues/200)
+    pub validator_addr: Address,
+    /// The new [`Epoch`].
+    ///
+    /// Since this is a monotonically growing sequence number,
+    /// it is signed together with the rest of the data to
+    /// prevent replay attacks on validator set updates.
+    pub epoch: Epoch,
+}
+
+impl Vext {
+    /// Creates a new signed [`Vext`].
+    ///
+    /// For more information, read the docs of [`SignedVext::new`].
+    #[inline]
+    pub fn sign(&self, sk: &common::SecretKey) -> SignedVext {
+        SignedVext::new_abi_encoded(sk, self.clone())
+    }
+}
+
+/// Provides a mapping between [`EthAddr`] and [`VotingPower`] instances.
+pub type VotingPowersMap = HashMap<EthAddr, VotingPower>;
+
+/// This trait contains additional methods for a [`HashMap`], related
+/// with validator set update vote extensions logic.
+pub trait VotingPowersMapExt {
+    /// Returns the keccak hash of this [`VotingPowersMap`]
+    /// to be signed by an Ethereum validator key.
+    fn get_bridge_hash(&self, epoch: Epoch) -> [u8; 32];
```

Comment:
A super nit, but I like using the `Hash` type instead unless there is a really good reason not to. I think it just make intentions clearer in the code.